Sprint #21 first try

nim0y opened the pull request:
//
AudioPlayer transfer to Fragment

nim0y committed:
Sprint #21 first try
//
AudioPlayer transfer to Fragment

nim0y committed:
Sprint #21 hot fix #1
//
showFavoritesList
correction

Nick-review commented on .idea/kotlinc.xml:
Папка .idea не должна попадать в пул реквесты, её надо занести в файл .gitignore
Это индивидуальные настройки вашей android studio

Nick-review commented on app/src/main/java/com/example/playlistmaker/data/db/TrackEntity.kt:
```diff
@@ -0,0 +1,21 @@
+data class TrackEntity(
+    @PrimaryKey
+    val trackId: String,
+    var trackName: String,
```
Почему какие-то свойства var, а какие-то val? вроде бы все можно сделать val

Nick-review commented on app/src/main/java/com/example/playlistmaker/data/db/TrackEntity.kt:
```diff
@@ -0,0 +1,21 @@
+import androidx.room.PrimaryKey
+import java.util.Date
+
+@Entity(tableName = "favorite_tracks_tablet")
```
Название таблиц можно вынести в список констант, это удобно если будут использовать в нескольких местах

Nick-review commented on app/src/main/java/com/example/playlistmaker/data/db/converters/TrackDbConverter.kt:
```diff
@@ -0,0 +1,40 @@
+        )
+    }
+
+    fun map(track: TrackEntity): Track {
```
Такие мапперы очень удобно писать как Extension

```
fun TrackEntity.toModel() = Track(
        trackId,
        trackName,
        artistName,
        trackTimeMillis,
        artworkUrl100,
        collectionName,
        releaseDate,
        primaryGenreName,
        country,
        previewUrl,
        insertTime
    )
```

Nick-review commented on app/src/main/java/com/example/playlistmaker/data/dto/TrackDto.kt:
```diff
@@ -1,14 +1,14 @@
-    val previewUrl: String?
+    val artworkUrl100: String,
+    val collectionName: String,
+    val releaseDate: String,
```
Лучше оставить нулабельность, в одной работе видел что не приходили некоторые поля

Nick-review commented on app/src/main/java/com/example/playlistmaker/data/impl/db/FavoriteTracksInteractorImpl.kt:
```diff
@@ -0,0 +1,25 @@
+package com.example.playlistmaker.data.impl.db
```
Неправильное расположение имплементации интерактора, он должен находиться в domain слое, и не в директории db
Грубо говоря, интерактор не должен знать откуда там берутся избранные треки, это уже ответственность репозитория

Nick-review commented on app/src/main/java/com/example/playlistmaker/di/DataModule.kt:
```diff
@@ -45,6 +47,12 @@ val dataModule = module {
     }
+
+    single {
+        Room.databaseBuilder(androidContext(), AppDatabase::class.java, "database.db")
```
название базы данных тоже можно в константу положить

Nick-review commented on app/src/main/java/com/example/playlistmaker/domain/db/FavoriteTracksInteractor.kt:
```diff
@@ -0,0 +1,14 @@
+
+    suspend fun deleteTrack(trackId: Int)
+
+    suspend fun isFavoriteTrack(trackId: Int): Flow<Boolean>
```
Зачем тут Flow? Можно вернуть просто Boolean

Nick-review commented on app/src/main/java/com/example/playlistmaker/domain/db/FavoriteTracksRepository.kt:
```diff
@@ -0,0 +1,14 @@
+
+    suspend fun deleteTrack(trackId: Int)
+
+    suspend fun isFavoriteTrack(trackId: Int): Flow<Boolean>
```
Аналогично с Flow, он тут не нужен

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/favorite/FavoriteState.kt:
```diff
@@ -0,0 +1,12 @@
+
+import com.example.playlistmaker.domain.models.search.Track
+
+sealed interface FavoriteState {
```
Хорошее решение, лайк

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/favorite/FavoriteTracksFragment.kt:
```diff
@@ -0,0 +1,101 @@
+
+    @SuppressLint("NotifyDataSetChanged")
+    private fun showFavoritesList(tracks: List<Track>) {
+        binding.favoriteListEmpty.visibility = View.GONE
```
with(binding) + управление видимостью лучше делать через isVisible = true/false как по мне

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/favorite/FavoriteTracksFragment.kt:
```diff
@@ -0,0 +1,101 @@
+        with(favoriteAdapter) {
+            tracksList.clear()
+            tracksList.addAll(tracks)
+            notifyItemRangeChanged(0, tracksList.size)
```
вот этот Notify лучше переносить сразу в адаптер в метод добавления треков, не надо будет каждый раз его дёргать в разных местах, а ведь можно и забыть его вызвать

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/favorite/FavoriteTracksFragment.kt:
```diff
@@ -0,0 +1,101 @@
+
+    override fun onResume() {
+        super.onResume()
+        vm.getFavoriteList()
```
Лучше в onViewCreated перенести, чтобы данные уже были подготовлены когда экран отрисуется, а то у вас получается что данные начинают загружаться когда экран уже отрисован - могут возникнуть задержки отрисовки

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/mediateka/favorite/FavoriteTracksViewModel.kt:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    init {
+        setState(FavoriteState.Load)
```
Можно сделать это начальным значением в _favoriteState и не надо будет писать блок init

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerFragment.kt:
```diff
@@ -2,49 +2,98 @@ package com.example.playlistmaker.ui.player
-        track = IntentCompat.getParcelableExtra(intent, SEARCH_QUERY_HISTORY, Track::class.java)!!
+        track = arguments?.getParcelable<Track>(SEARCH_QUERY_HISTORY)!!
+
+        (activity as MainActivity).hideNavBar()
```
такие касты лучше делать через as?

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerFragment.kt:
```diff
@@ -2,49 +2,98 @@ package com.example.playlistmaker.ui.player
-        vm.audioPlayerState.observe(this) { state ->
+        vm.isFavorite(track)
+
+        vm.audioPlayerState.observe(viewLifecycleOwner) { state ->
```
как правило, подписки выносятся в отдельные метод, по типу render state

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerFragment.kt:
```diff
@@ -2,49 +2,98 @@ package com.example.playlistmaker.ui.player
+
+    override fun onDestroyView() {
+        super.onDestroyView()
+        vm.onDestroy()
```
у вью модели есть свой метод ЖЦ, называется onCleared, эту логику в дестроем вью модели надо туда

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerViewModel.kt:
```diff
@@ -14,16 +17,21 @@ import kotlinx.coroutines.launch
 ) : ViewModel() {
 
+    private val _favoriteState = MutableLiveData<FavoriteState>()
+    val favoriteState: LiveData<FavoriteState> = _favoriteState
```
почему favorite state? тогда player state или просто state

Nick-review commented on app/src/main/res/layout/fragment_favorite_tracks.xml:
```diff
@@ -1,31 +1,53 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center"
+            android:layout_margin="16dp"
```
размеры надо вынести в dimens.xml

Nick-review commented on app/src/main/res/layout/fragment_favorite_tracks.xml:
```diff
@@ -1,31 +1,53 @@
-        android:layout_height="wrap_content"
+    <com.google.android.material.progressindicator.CircularProgressIndicator
+        android:id="@+id/favorites_progress_bar"
+        android:layout_width="44dp"
```
размеры надо вынести в dimens.xml

nim0y committed:
Sprint #21 try #2
//
Никита , спасибо за ревью!

У наставников спросил, пока жду ответа.(морозятся что то ...) Спрошу и у тебя, если есть возможность ответь, пожалуйста. (коли нет, то не страшно)

 Похоже от меня ускользает смысл этого пункта.

Если можно по подробнее(схемку аль чертёж) напиши про :
В реализации репозитория, осуществляющего поиск треков, используется метод получения из БД списка идентификаторов треков, добавленных в избранное.

С уважением,
Михаил.

nim0y committed:
Sprint #21 little commit
//

Nick-review commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerViewModel.kt:
```diff
@@ -55,7 +55,7 @@ class AudioPlayerViewModel(
         viewModelScope.launch {
             favoriteTracksInteractor
-                .isFavoriteTrack(track.trackId.toInt())
+                .isFavoriteTrack(track.trackId!!)
```
в таких случая лучше использовать .orZero() для Int или .orEmpty() для String, или ?: с дефолтным значение, т.к. в твоём случае приложение упадёт с NullPointerException в случае чего

nim0y committed:
Sprint #21 tiny commit
